### ISSUE TITLE
getting loss as nan on rocm enabled system

### ISSUE THREAD

**Issue by PattasuBalu** [pytorch/pytorch#110532]
### 🐛 Describe the bug

#transformations
train_transforms = transforms.Compose([transforms.Resize((224,224)),
                                       transforms.ToTensor(),                                
                                       torchvision.transforms.Normalize(
                                           mean=[0.485, 0.456, 0.406],
                                           std=[0.229, 0.224, 0.225],
    ),
                                       ])
test_transforms = transforms.Compose([transforms.Resize((224,224)),
                                      transforms.ToTensor(),
                                      torchvision.transforms.Normalize(
                                          mean=[0.485, 0.456, 0.406],
                                          std=[0.229, 0.224, 0.225],
    ),
                                      ])

#datasets
train_data = datasets.ImageFolder(traindir,transform=train_transforms)
test_data = datasets.ImageFolder(testdir,transform=test_transforms)

#dataloader
trainloader = torch.utils.data.DataLoader(train_data, shuffle = True, batch_size=16)
testloader = torch.utils.data.DataLoader(test_data, shuffle = True, batch_size=16)
print(len(trainloader))
print(len(testloader))

def make_train_step(model, optimizer, loss_fn):
  def train_step(x,y):
    #make prediction
    yhat = model(x)
    #enter train mode
    model.train()
    #compute loss
    loss = loss_fn(yhat,y)

    loss.backward()
    optimizer.step()
    optimizer.zero_grad()
    #optimizer.cleargrads()

    return loss
  return train_step

from torchvision import datasets, models, transforms
import torch.nn as nn

device = "cuda"
model = models.resnet18(pretrained=True)

#freeze all params
for params in model.parameters():
  params.requires_grad_ = False

#add a new final layer
nr_filters = model.fc.in_features  #number of input features of last layer
model.fc = nn.Linear(nr_filters, 1)

model = model.to(device)
from torch.nn.modules.loss import BCEWithLogitsLoss
from torch.optim import lr_scheduler

#loss
loss_fn = BCEWithLogitsLoss() #binary cross entropy with sigmoid, so no need to use sigmoid in the model

#optimizer
optimizer = torch.optim.Adam(model.fc.parameters()) 

#train step
train_step = make_train_step(model, optimizer, loss_fn)

!pip install tqdm
from tqdm import tqdm

losses = []
val_losses = []

epoch_train_losses = []
epoch_test_losses = []

n_epochs = 2
early_stopping_tolerance = 3
early_stopping_threshold = 0.03

for epoch in range(n_epochs):
  epoch_loss = 0
  for i ,data in tqdm(enumerate(trainloader), total = len(trainloader)): #iterate ove batches
    x_batch , y_batch = data
    x_batch = x_batch.to(device) #move to gpu
    y_batch = y_batch.unsqueeze(1).float() #convert target to same nn output shape
    y_batch = y_batch.to(device) #move to gpu

    loss = train_step(x_batch, y_batch)
    epoch_loss += loss/len(trainloader)
    losses.append(loss)
    
  epoch_train_losses.append(epoch_loss)
  print('\nEpoch : {}, train loss : {}'.format(epoch+1,epoch_loss))

  #validation doesnt requires gradient
  with torch.no_grad():
    cum_loss = 0
    for x_batch, y_batch in testloader:
      x_batch = x_batch.to(device)
      y_batch = y_batch.unsqueeze(1).float() #convert target to same nn output shape
      y_batch = y_batch.to(device)

      #model to eval mode
      model.eval()

      yhat = model(x_batch)
      val_loss = loss_fn(yhat,y_batch)
      cum_loss += loss/len(testloader)
      val_losses.append(val_loss.item())

    epoch_test_losses.append(cum_loss)
    print('Epoch : {}, val loss : {}'.format(epoch+1,cum_loss))  
    
    best_loss = min(epoch_test_losses)
    
    #save best model
    if cum_loss <= best_loss:
      best_model_wts = model.state_dict()
    
    #early stopping
    early_stopping_counter = 0
    if cum_loss > best_loss:
      early_stopping_counter +=1

    if (early_stopping_counter == early_stopping_tolerance) or (best_loss <= early_stopping_threshold):
      print("/nTerminating: early stopping")
      break #terminate training
    
#load best model
model.load_state_dict(best_model_wts)

100%|█████████████████████████████████████████| 501/501 [00:26<00:00, 19.26it/s]

Epoch : 1, train loss : nan
Epoch : 1, val loss : nan
100%|█████████████████████████████████████████| 501/501 [00:24<00:00, 20.58it/s]

Epoch : 2, train loss : nan
Epoch : 2, val loss : nan
100%|█████████████████████████████████████████| 501/501 [00:23<00:00, 21.60it/s]

The issue seems to be with torch as the code gets working on a google colab notebook with cuda.

### Versions

2.0.1+rocm5.4.2

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

**Comment by z-a-f** [comment]
I think you will have faster response if you post this on the [forum](https://discuss.pytorch.org/).

As for the problem itself, you need to check:

1. Is your learning rate too big? If so, make it smaller
2. Are your gradients really large? If so, clip them
3. Are your inputs valid? If not, clean them up

On other notes:

1. Call the `zero_grad` before calling the model -- there are several moving parts, I didn't go into details, but just in case
2. You are setting the best model to the state_dict. I am not 100% sure, but I think it's mutable, so it will be replaced on the next iteration

**Comment by PattasuBalu** [comment]
The model works fine on Google Colab with same code which I have mentioned above. The issue seems to be with Pytorch. The same issue I have encountered is also mentioned in the [https://discuss.pytorch.org/t/sending-data-to-device-in-pytorch-causes-it-to-contain-nan-values-rocm/177711](forum). Any fix this issue would be of great help.

**Comment by hongxiayang** [comment]
Can you provide the steps and simple code so that it is reproducible? It seems your description was not formatted/indented well. 

Also, please let me know whether this works fine or not:

 python -c "import torch; x = torch.ones(1,2).cuda(); print(x); print('sum {}'.format(torch.sum(x)))"

**Comment by hongxiayang** [comment]
The wrong output for the above test indicated an issue related to PCIe atomics in your computer. It is a known issue. For now, either you use an older version of rocm (like 5.2), or upgrading your cpu/motherboard as indicated in the latest documentation:
![image](https://github.com/pytorch/pytorch/assets/62075498/75ac9abd-48b0-43af-8f25-dc4158a8a52b)

We have been discussing this in length in this issue https://github.com/pytorch/pytorch/issues/103973.

**Comment by PattasuBalu** [comment]
Ok. But how do I downgrade my rocm version to 5.2?

**Comment by hongxiayang** [comment]
> Ok. But how do I downgrade my rocm version to 5.2?

You can test this docker image which has the rocm 5.2, though it is pretty old. You can check whether this works for you.
```
rocm/pytorch:rocm5.2.3_ubuntu20.04_py3.7_pytorch_1.11.0
```

Add this alias:
```
alias drun='sudo docker run -it --network=host --device=/dev/kfd --device=/dev/dri --group-add=video --ipc=host --cap-add=SYS_PTRACE --security-opt seccomp=unconfined --shm-size 8G -v $HOME/dockerx:/dockerx -w /dockerx'
```

then 
```
drun rocm/pytorch:rocm5.2.3_ubuntu20.04_py3.7_pytorch_1.11.0
```

Inside the container, run the test. It is expected to see an output of 2.0:
```
python3 -c "import torch; x = torch.ones(1,2).cuda(); print(x); print('sum {}'.format(torch.sum(x)))"
```

**Comment by PattasuBalu** [comment]
The above mentioned docker image is not working. Is there anything else I can do to fix this?

Digest: sha256:23cacf07d342ff85ddfe226a083743ff81362eceef9ad8d0e2f695f1213c3e12
Status: Downloaded newer image for rocm/pytorch:rocm5.2.3_ubuntu20.04_py3.7_pytorch_1.11.0
docker.io/rocm/pytorch:rocm5.2.3_ubuntu20.04_py3.7_pytorch_1.11.0

**Comment by hongxiayang** [comment]
What error did you get?

**Comment by PattasuBalu** [comment]
The docker image is not downloading completely and i got the message mentioned in the above message telling to download a newer version.

**Comment by PattasuBalu** [comment]
The issue is now fixed after installing torch 1.13.0+rocm5.2. The above mentioned command too runs perfectly without any issues. I did not downgrade rocm version to 5.2. I am still running rocm 5.4.2, but torch works fine. Thank you for your help @hongxiayang.
The result of the command:

(torch) pc@pc:~/test$ python3 -c "import torch; x = torch.ones(1,2).cuda(); print(x); print('sum {}'.format(torch.sum(x)))"
tensor([[1., 1.]], device='cuda:0')
sum 2.0

**Comment by hongxiayang** [comment]
@PattasuBalu Great, glad to hear you made it work.

**Comment by PattasuBalu** [comment]
Do I need to close this issue or what?😅

**Comment by hongxiayang** [comment]
yeah, please close it. We have other issues that we discuss atomics. So, we don't need to duplicate.
